server/leafnode.go

```diff
@@ -1068,6 +1071,42 @@ func (c *client) processLeafNodeConnect(s *Server, arg []byte, lang string) erro
 	return nil
 }
 
+// Invoked from a server accepting a leafnode connection. It looks for a possible
+// existing leafnode connection from the same server with the same account, and
+// if it finds one, closes it so that the new one is accepted and not reported as
+// forming a cycle.
+//
+// This must be invoked for LEAF client types, and on the server accepting the connection.
+//
+// No server nor client lock held on entry.
+func (c *client) replaceOldLeafNodeConnIfNeeded(s *Server, connInfo *leafConnectInfo) {
+	var accName string
+	c.mu.Lock()
+	if c.acc != nil {
+		accName = c.acc.Name
+	}
+	c.mu.Unlock()
+
+	var old *client
+	s.mu.Lock()
+	for _, ol := range s.leafs {
+		ol.mu.Lock()
+		// We check for empty because in some test we may send empty CONNECT{}
+		if ol.opts.Name == connInfo.Name && connInfo.Name != _EMPTY_ && ol.acc.Name == accName {
+			old = ol
+		}
+		ol.mu.Unlock()
+		if old != nil {
+			break
+		}
+	}
+	s.mu.Unlock()
+	if old != nil {
+		old.Warnf("Replacing connection from same server")
+		old.closeConnection(ReadError)
```

Comment:
would StaleConnection be a more appropriate closedState?

Reply:
I went from the principle that this is what would be reported in a normal disconnect, in the read loop, when getting an error on Read(). But I can change it if you think it makes more sense. (to me, stale is used when we close connection after sending max out pings without corresponding pongs).

Reply:
ok, that makes sense. LGTM